alternator/executor.cc

```diff
@@ -1636,19 +1649,56 @@ future<executor::request_return_type> rmw_operation::execute(service::storage_pr
     });
 }
 
+bytes query_cache_key(const std::string_view& query, char query_type) {
+   auto b = md5_hasher::calculate(query);
+   b[0] = query_type; // this avoids key collision for different expression types
+   return b;
+}
+
+template<typename T, typename F>
+requires requires(F f, std::string_view v) {
+    { f(v) } -> std::same_as<T>;
+}
+static T parse_or_get_from_cache(lw_shared_ptr<expression_cache_t> cache, F parse_func, std::string_view query, char query_type) {
+    auto key = query_cache_key(query, query_type);
+    auto ptr = cache->find(key);
+    if (ptr) {
+        // we need to copy, it can be evicted from cache anytime and
+        // aditonally our expression evaluation code modifies T
+        return std::get<T>(*ptr);
+    }
+    // not found in cache, we need to put it there
+    auto loader_func = [parse_func, query_copy = std::string(query)] (bytes key) {
+        return make_ready_future<expression_cache_value_t>(parse_func(query_copy));
+    };
+    // leak future, we have no means to wait for it here
+    (void)cache->get_ptr(key, loader_func).discard_result()
+        .handle_exception([](auto ep) {/* ignore */}).then_wrapped([cache] (auto f) {
+        (void)cache; // keep it alive until we're done with it
+        return make_ready_future();
+    });
+    // parse it second time as synchronization with background task is not possible here
+    return parse_func(query);
+}
+
 parsed::condition_expression
 executor::parse_condition_expression(std::string_view query, const char* caller) {
-    return parsed::parse_condition_expression(query, caller);
+    return parse_or_get_from_cache<parsed::condition_expression>(_expression_cache,
+        [caller] (std::string_view query) {
+            return parsed::parse_condition_expression(query, caller);
+        }, query, 'C');
 }
```

Comment:
Unrelated: we should move cql3::expr::expression to expr::expression and convert alternator conditions (and other expressions) to general expressions.

Reply:
I don't think that's a good idea. Alternator expressions have different features, semantics, error-handling and weird edge-cases, from CQL. If we try to unify them we'll need to add a lot of ugly options that will make what you had to do with those "null handling" options pale.

Alternator's expressions are done, and already do what they should. Perhaps in the future we'll need some way to send them over Raft or offload them to a replica, which we can't do now, but until we really want to do those things, I don't think we need to rewrite them.

Reply:
Expressions are expressions, they have inputs and outputs. If we send alternator expressions on raft, we'll want to send cql expressions on raft.

Reply:
In theory, yes, we could have a virtual machine that can run any expressions. But it practice, the current CQL and DynamoDB expressions have different data types, different operations, different error messages and even error cases - we can't just translate DynamoDB expressions to CQL expressions, although we could eventually invent one grand-unified-expression language that will allow both.

As I noted above, you've already begun to see those problems when unifying the expressions for CQL "where" and "if" clauses. Although these expressions are almost identical, you had difference in NULL handling, which made you need to add "configuration options" for these expressions on whether to behave like "where" or "if". We'll need to add more of these options - as well as more types and more operations - to support Alternator in the same expression tree.

Reply:
I'm fine with adding options. It's better than having two different implementations of the same thing.

Reply:
> I'm fine with adding options. It's better than having two different implementations of the same thing.

I theory, I agree. In practice, there will be almost nothing in common between the two types of expressions (even when you have the same operator like "<", they will have different error handling, different types, and so on), and many operations and functions will be unique to just one of the two languages. We can do this eventually, but I don't think it's related to this PR - or urgent.